pkgs/build-support/nix-gitignore/default.nix

```diff
@@ -147,10 +147,10 @@ in rec {
         ' sh {} \; > $out
       '');
 
-  withGitignoreFile = patterns: root:
-    lib.toList patterns ++ [(root + "/.gitignore")];
+  withGitIgnoredFilesAndDirs = patterns: root:
+    lib.toList patterns ++ [(root + "/.gitignore") (".git")];
```

Comment:
Shouldn’t this be `(root + "/.git")`? This looks like we are ignoring everything that contains `.git` really.

Reply:
no, it will try to read `.git` dir

https://github.com/siers/nix-gitignore/blob/master/default.nix#L97-L98

and throw with error

```
 ~/projects/nix-gitignore   master  ./run-tests.sh
nix-gitignore: using test dir /tmp/tmp.eSD2CjC4hr
error (ignored): writing to file: Bad file descriptor
error: while evaluating the attribute 'success' at /home/srghma/projects/nix-gitignore/test.nix:207:3:
while evaluating 'sourceActualAux' at /home/srghma/projects/nix-gitignore/test.nix:154:21, called from /home/srghma/projects/nix-gitignore/test.nix:159:36:
while evaluating 'gitignoreFilterSource' at /home/srghma/projects/nix-gitignore/default.nix:159:45, called from /home/srghma/projects/nix-gitignore/test.nix:154:26:
while evaluating 'gitignoreFilterSourcePure' at /home/srghma/projects/nix-gitignore/default.nix:156:49, called from /home/srghma/projects/nix-gitignore/default.nix:160:5:
while evaluating 'gitignoreFilterPure' at /home/srghma/projects/nix-gitignore/default.nix:101:55, called from undefined position:
while evaluating 'filterPattern' at /home/srghma/projects/nix-gitignore/default.nix:21:12, called from /home/srghma/projects/nix-gitignore/default.nix:102:5:
while evaluating 'last' at /home/srghma/projects/nix-gitignore/default.nix:12:10, called from /home/srghma/projects/nix-gitignore/default.nix:27:9:
while evaluating 'last' at /home/srghma/projects/nix-gitignore/default.nix:12:10, called from /home/srghma/projects/nix-gitignore/default.nix:27:15:
while evaluating 'gitignoreToPatterns' at /home/srghma/projects/nix-gitignore/default.nix:31:25, called from /home/srghma/projects/nix-gitignore/default.nix:92:47:
while evaluating 'gitignoreCompileIgnore' at /home/srghma/projects/nix-gitignore/default.nix:95:47, called from /home/srghma/projects/nix-gitignore/default.nix:102:22:
while evaluating 'onPath' at /home/srghma/projects/nix-gitignore/default.nix:97:19, called from undefined position:
opening file '/tmp/tmp.eSD2CjC4hr/test-tree-normal/.git': No such file or directory
```

I've added tests https://github.com/siers/nix-gitignore/pull/25

Reply:
So does that mean we can change the `root + "/.gitignore"` to `".gitignore"`?

Reply:
ping ;)

Reply:
no, it fails tests, dont know why

```sh
building '/nix/store/bkzay1c3iw2m44mf4x7id0mcyy33q9lh-nix-gitignore-test.drv'...
2d1
< .gitignore
5a5
> 1-simpl/1
24a25,26
> 1-simpl/1-simpl/simpletest
> 1-simpl/2
29a32
> 1-simpl/\
31a35
> 1-simpl/]]
35a40
> 1-simpl/bb
36a42
> 1-simpl/simple-test
43a50
> 2-negation/20
44a52
> 2-negation/40
45a54
> 2-negation/60
47a57
> 3-wildcards/bar.html
49a60
> 3-wildcards/foo.html
50a62
> 3-wildcards/html/bar.html
54a67,69
> 4-escapes/*.html
> 4-escapes/o?ther.html
> 4-escapes/o__ther.html
56a72
> 4-escapes/other.html$
57a74,92
> 5-directory
> 5-directory/$
> 5-directory/$^
> 5-directory/1
> 5-directory/2
> 5-directory/3
> 5-directory/4
> 5-directory/5
> 5-directory/[
> 5-directory/[[
> 5-directory/\
> 5-directory/\\
> 5-directory/]
> 5-directory/]]
> 5-directory/]]]
> 5-directory/^
> 5-directory/^$
> 5-directory/ab
> 5-directory/bb
builder for '/nix/store/bkzay1c3iw2m44mf4x7id0mcyy33q9lh-nix-gitignore-test.drv' failed with exit code 1
error: build of '/nix/store/bkzay1c3iw2m44mf4x7id0mcyy33q9lh-nix-gitignore-test.drv' failed
```

Reply:
but without this change it works

Reply:
@Profpatsch I think we can merge

Reply:
I have to take another closer look on Monday or Tuesday.

Reply:
It fails because `root` is a path and so is `root + "/.git".

Now, I added this feature that *reads* path values and uses the ignore patterns in them. Because this, unlike, `root + "/.gitignore"` should be a literal pattern, we have to cast it to string with `(toString (root + "/.git"))`.

Reply:
Nah, I am wrong! `root + ` should be prepended so it can be read as a file of patterns. `".git"`, however, is a pattern, so it shouldn't be a full path.

Reply:
It's going to be a little tricky with the testcases because of 1. `shopt -s dotglob` in `sourceGitFrom`, 2. because `sourceGitFrom` does `rm -r .git` before handing over the filtered contents.

I didn't succeed to create them yet.

```diff --git a/default.nix b/default.nix
index b0bf3ca..45aa1c4 100644
--- a/default.nix
+++ b/default.nix
@@ -146,10 +146,10 @@ in rec {
       '');

   withGitIgnoredFilesAndDirs = patterns: root:
-    lib.toList patterns ++ [(root + "/.gitignore") (".git")];
+    lib.toList patterns ++ [(root + "/.gitignore") ".git"];

   withRecursiveGitIgnoredFilesAndDirs = patterns: root:
-    lib.toList patterns ++ [(compileRecursiveGitignore root)];
+    lib.toList patterns ++ [(compileRecursiveGitignore root) ".git"];

   # filterSource derivatives
```

This is how the `default.nix` code should look though.